stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/ConnectionAction.java

```diff
@@ -61,15 +64,52 @@
                                                 BStruct outboundResponseStruct, HTTPCarbonMessage responseMessage) {
-        BStruct entityStruct = MimeUtil.extractEntity(outboundResponseStruct);
+        String contentType = HttpUtil.getContentTypeFromTransportMessage(responseMessage);
+        String boundaryString = null;
+        if (HeaderUtil.isMultipart(contentType)) {
+            boundaryString = HttpUtil.addBoundaryIfNotExist(responseMessage, contentType);
+        }
         HttpResponseFuture outboundRespStatusFuture = HttpUtil.sendOutboundResponse(requestMessage, responseMessage);
+        BStruct entityStruct = MimeUtil.extractEntity(outboundResponseStruct);
         if (entityStruct != null) {
-            MessageDataSource outboundMessageSource = EntityBodyHandler.getMessageDataSource(entityStruct);
-            serializeMsgDataSource(responseMessage, outboundMessageSource, outboundRespStatusFuture, entityStruct);
+            if (boundaryString != null) {
+                serializeMultiparts(responseMessage, boundaryString, outboundRespStatusFuture, entityStruct);
+            } else {
+                MessageDataSource outboundMessageSource = EntityBodyHandler.getMessageDataSource(entityStruct);
+                serializeMsgDataSource(responseMessage, outboundMessageSource, outboundRespStatusFuture, entityStruct);
+            }
         }
         return handleResponseStatus(context, outboundRespStatusFuture);
     }
 
+    /**
+     * Serlaize multipart entity body. If an array of body parts exist, encode body parts else serialize body content
+     * if it exist as a byte channel.
+     *
+     * @param responseMessage          Response message that needs to be sent out.
+     * @param boundaryString           Boundary string that should be used in encoding body parts
+     * @param outboundRespStatusFuture Represent the future events and results of connectors
+     * @param entityStruct             Represent the entity that holds the actual body
+     */
+    private void serializeMultiparts(HTTPCarbonMessage responseMessage, String boundaryString,
+                                     HttpResponseFuture outboundRespStatusFuture, BStruct entityStruct) {
+        BRefValueArray bodyParts = EntityBodyHandler.getBodyPartArray(entityStruct);
+        if (bodyParts != null && bodyParts.size() > 0) {
+            MultipartDataSource multipartDataSource = new MultipartDataSource(entityStruct, boundaryString);
+            serializeMsgDataSource(responseMessage, multipartDataSource, outboundRespStatusFuture,
+                    entityStruct);
+        } else {
+            OutputStream messageOutputStream = getOutputStream(responseMessage, outboundRespStatusFuture);
+            try {
+                EntityBodyHandler.writeByteChannelToOutputStream(entityStruct, messageOutputStream);
+                HttpUtil.closeMessageOutputStream(messageOutputStream);
```

Comment:
If an exception occur before this, seems output stream would not be closed. Can we use finally clause here?